File: PolskiWhisper/Sources/Features/Dictation/WhisperHallucinationFilter.swift
//
//  WhisperHallucinationFilter.swift
//  PolskiWhisper
//
//  Copyright © 2026 Marcin Werner. Licensed under the MIT License.
//  See LICENSE in the repository root.
//

import Foundation

/// Filtruje znane halucynacje Whispera - frazy które model wstawia gdy dostaje
/// cichy lub niewyraźny audio (trenowany na YouTube subtitles z outros).
///
/// Strategia:
/// 1. Jeśli **cała** transkrypcja matchuje halucynację → zwróć pusty string
/// 2. Jeśli halucynacja jest **fragmentem** (na początku/końcu) → usuń ten fragment
/// 3. Inne pozostawiamy (mogą być legitne wystąpienia w środku zdania)
@MainActor
enum WhisperHallucinationFilter {

    /// Lista typowych halucynacji Whispera w języku polskim i angielskim.
    /// Case-insensitive matching.
    private static let knownHallucinations: [String] = [
        // Polish YouTube outros
        "Dzięki za oglądanie",
        "Dziękuję za oglądanie",
        "Dzięki za uwagę",
        "Dziękuję za uwagę",
        "Subskrybujcie",
        "Lubcie i subskrybujcie",
        "Subskrybujcie kanał",
        "Wszystkie prawa zastrzeżone",
        "Do zobaczenia",
        "Do zobaczenia w następnym odcinku",
        "Do następnego",
        "Trzymajcie się",
        "Pa pa",
        "Napisy stworzone przez społeczność Amara.org",
        "Napisy stworzone przez",
        "Tłumaczenie:",
        "Korekta:",

        // English (czasem leci nawet z polish lang setting)
        "Thanks for watching",
        "Thank you for watching",
        "Thanks for watching!",
        "Subscribe to my channel",
        "Don't forget to subscribe",
        "Like and subscribe",
        "See you next time",
        "Bye bye",
        "All rights reserved",
        "Subtitles by the Amara.org community",

        // Bardzo krótkie (~ pojedyncze słowa-halucynacje)
        "you",
        "Bye",
        "Thanks"
    ]

    /// Filtruje halucynacje z transkrypcji.
    static func filter(_ text: String) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "" }

        // 1. Cała transkrypcja matchuje halucynację? → wyrzuć całość
        for hallucination in knownHallucinations {
            if textEqualsIgnoringPunctuation(trimmed, hallucination) {
                return ""
            }
        }

        // 2. Sprawdź czy transkrypcja składa się WYŁĄCZNIE z halucynacji
        // (kilka połączonych typu "Dzięki za oglądanie. Subskrybujcie!")
        if isOnlyHallucinations(trimmed) {
            return ""
        }

        // 3. Usuń halucynacje na końcu zdania (zachowaj resztę)
        var result = trimmed
        for hallucination in knownHallucinations {
            // Match at end z opcjonalnymi znakami końcowymi
            let escaped = NSRegularExpression.escapedPattern(for: hallucination)
            let pattern = "\\s*\\.?\\s*\(escaped)\\s*[!.?]*\\s*$"
            if let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) {
                let range = NSRange(result.startIndex..., in: result)
                result = regex.stringByReplacingMatches(
                    in: result,
                    options: [],
                    range: range,
                    withTemplate: ""
                ).trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }

        return result
    }

    // MARK: - Private helpers

    /// Strip polskich znaków diakrytycznych. Standard `.stripDiacritics` zhandle
    /// ą→a, ę→e, ć→c, ń→n, ó→o, ś→s, ź→z, ż→z, ALE NIE ł→l (bo `ł` to osobny
    /// Unicode glyph U+0142, nie litera+diakrytyk). Dlatego manual replacement.
    private static func foldPolish(_ text: String) -> String {
        let stripped = text.applyingTransform(.stripDiacritics, reverse: false) ?? text
        return stripped
            .replacingOccurrences(of: "ł", with: "l")
            .replacingOccurrences(of: "Ł", with: "L")
    }

    /// Porównuje dwa stringi ignorując punktuację, case, whitespace **i polskie ogonki**.
    /// Diacritic-fold łapie warianty typu "Dziekuje za ogladanie" gdy Whisper zgubi polskie znaki.
    private static func textEqualsIgnoringPunctuation(_ a: String, _ b: String) -> Bool {
        let cleaned: (String) -> String = { text in
            foldPolish(text.lowercased())
                .components(separatedBy: CharacterSet.alphanumerics.union(.whitespaces).inverted)
                .joined()
                .trimmingCharacters(in: .whitespacesAndNewlines)
                .replacingOccurrences(of: "  ", with: " ")
        }
        return cleaned(a) == cleaned(b)
    }

    /// Sprawdza czy text składa się WYŁĄCZNIE z halucynacji (np. "Dzięki za oglądanie. Bye bye.").
    /// Algorytm: po usunięciu wszystkich halucynacji nie zostaje nic znaczącego.
    /// Strip diacritics + ł przed porównaniem - łapie warianty bez polskich znaków.
    private static func isOnlyHallucinations(_ text: String) -> Bool {
        var stripped = foldPolish(text.lowercased())
        for hallucination in knownHallucinations {
            let folded = foldPolish(hallucination.lowercased())
            stripped = stripped.replacingOccurrences(of: folded, with: "")
        }
        // Po usunięciu zostały tylko białe znaki + punktuacja?
        let alphanumericLeft = stripped.unicodeScalars
            .filter { CharacterSet.alphanumerics.contains($0) }
            .count
        return alphanumericLeft < 3  // mniej niż 3 znaków = praktycznie nic
    }
}

File: windows/src/PolskiWhisperWin.Core/Services/WhisperHallucinationFilter.cs
// PolskiWhisperWin
// Copyright © 2026 Marcin Werner. Licensed under the MIT License.
// See LICENSE in the repository root.

using System.Globalization;
using System.Text;
using Microsoft.Extensions.Logging;

namespace PolskiWhisperWin.Core.Services;

/// <summary>
/// Filtruje typowe halucynacje Whisper przy ciszy lub szumie tła.
/// Mapping 1:1 do macOS <c>WhisperHallucinationFilter</c> (v0.1.1+).
/// </summary>
/// <remarks>
/// Whisper często zwraca "Dziękuję za oglądanie", "Napisy stworzone przez społeczność Amara.org"
/// i podobne fragmenty napisów Youtube przy bardzo cichym audio. Filter łapie:
/// 1) Dokładne dopasowanie znanych fraz
/// 2) Diacritic-fold porównanie (łapie warianty bez polskich znaków)
/// 3) Bardzo krótkie wyniki (1-2 znaki) - prawdopodobnie noise
/// </remarks>
public sealed class WhisperHallucinationFilter
{
    private readonly ILogger<WhisperHallucinationFilter> _logger;

    /// <summary>
    /// Znane frazy halucynowane przez Whisper przy ciszy. Wszystkie w lowercase
    /// + diacritic-folded (bez polskich znaków) dla case-insensitive matching.
    /// </summary>
    private static readonly IReadOnlySet<string> KnownHallucinations = new HashSet<string>(StringComparer.Ordinal)
    {
        // Polskie napisy YouTube hallucinations
        "dziekuje za ogladanie",
        "dziekuje za uwage",
        "dziekuje za obejrzenie filmu",
        "napisy stworzone przez spolecznosc amara.org",
        "napisy: amara.org",
        "amara.org",
        "subskrybuj kanal",
        "lajkuj i subskrybuj",
        "do zobaczenia w nastepnym filmie",
        "do zobaczenia",
        // English/multilingual hallucinations
        "thanks for watching",
        "thank you for watching",
        "thank you",
        "subscribe",
        "like and subscribe",
        ".",
        "...",
        "you",
        "[music]",
        "[muzyka]",
        "(muzyka)",
        "(music)",
    };

    public WhisperHallucinationFilter(ILogger<WhisperHallucinationFilter> logger)
    {
        _logger = logger;
    }

    /// <summary>
    /// Sprawdza czy <paramref name="text"/> wygląda na halucynację.
    /// Zwraca <c>true</c> jeśli tekst powinien zostać odrzucony (NIE wklejać).
    /// </summary>
    public bool IsLikelyHallucination(string? text)
    {
        if (string.IsNullOrWhiteSpace(text)) return true;

        var trimmed = text.Trim();

        // Bardzo krótki wynik (1-2 znaki) - prawdopodobnie noise.
        if (trimmed.Length <= 2)
        {
            _logger.LogDebug("Hallucination filter: tekst za krótki ({Length} znaków).", trimmed.Length);
            return true;
        }

        // Dokładne dopasowanie do znanych fraz (po normalizacji).
        var normalized = NormalizeForComparison(trimmed);
        if (KnownHallucinations.Contains(normalized))
        {
            _logger.LogInformation("Hallucination filter: dopasowano znaną frazę '{Phrase}'.", normalized);
            return true;
        }

        return false;
    }

    /// <summary>
    /// Normalizuje tekst do porównania: lowercase, trim, strip polish diacritics.
    /// "Dziękuję za oglądanie" → "dziekuje za ogladanie".
    /// </summary>
    public static string NormalizeForComparison(string input)
    {
        if (string.IsNullOrEmpty(input)) return input;

        var lowered = input.ToLowerInvariant().Trim();

        // Diacritic-fold via NFD + filter combining marks.
        var normalizedToFormD = lowered.Normalize(NormalizationForm.FormD);
        var sb = new StringBuilder(normalizedToFormD.Length);
        foreach (var ch in normalizedToFormD)
        {
            var category = CharUnicodeInfo.GetUnicodeCategory(ch);
            if (category != UnicodeCategory.NonSpacingMark)
                sb.Append(ch);
        }

        // Polskie "ł" nie jest combining mark - manualne mapping.
        return sb.ToString().Normalize(NormalizationForm.FormC)
            .Replace('ł', 'l')
            .Replace('Ł', 'l');
    }
}
